### PR TITLE
Optionally allow additional configurable characters in L057

### PR DESCRIPTION
**Pull request by tunetheweb**
### Brief summary of the change made

Allow hyphens in L057 as quoted identifiers (same as we allow underscores), as a lot of these are flagging for a lot of my table names in BigQuery.

@jpers36 , any issues with this since you created L057? If you would prefer not to allow this, then happy to move to a BigQuery specific bit of L057 since that already exists.

### Are there any other side effects of this change that we should be aware of?
None

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

**Comment by tunetheweb** [comment]
Can we hold off merging this until @jpers36 gives his opinion please?

**Comment by jpers36** [comment]
@tunetheweb maybe this is something that would best be handled in a config setting, such as either a flag or, more flexibly, a list of allowable non-alphanumerics.

**Comment by tunetheweb** [comment]
> @tunetheweb maybe this is something that would best be handled in a config setting, such as either a flag or, more flexibly, a list of allowable non-alphanumerics.

Good idea done.

In theory, this would allow us to get rid of `allow_space_in_identifier` and handle that in that. Alternatively could keep that (for backwards compatibility and for ease of setting as not always obvious should use quotes). WDYT?

**Comment by jpers36** [comment]
@tunetheweb makes sense to me to leave `allow_space_in_identifier ` for now for backwards compatibility.  We can always remove it later.

**Comment by tunetheweb** [comment]
OK think this is ready for review now.

File: src/sqlfluff/rules/L057.py
Comment (barrywhart) [inline src/sqlfluff/rules/L057.py:68]: Nitpick: IIUC, we are ignoring occurrences of `_` as well as `additional_allowed_identifiers`, but doing so in two different ways. Think it might be clearer to incorporate the `_` into the other code somehow? One thought: set the default value of `additional_allowed_identifiers` to `_`. 

Also, it looks like this same (increasingly complex) code is repeated 3 times in the rule. Think we could avoid this duplication?

File: src/sqlfluff/rules/L057.py
Comment (barrywhart) [inline src/sqlfluff/rules/L057.py:57]: These aren't really "identifiers". Rather, they are additional _characters_ allowed in identifiers. Perhaps call it `additional_allowed_identifier_characters` for clarity?

**Review by barrywhart — CHANGES_REQUESTED** [review]

File: src/sqlfluff/rules/L057.py
Comment (tunetheweb) [reply src/sqlfluff/rules/L057.py]: I think I'd prefer to keep the underscores in there. To me they should always be allowed. Petty Python doesn't have an `isalnum` equivalent that include them.

Also means we can skip the slightly more complex additional logic if it's not set (which it isn't by default) which would be a slight performance gain... maybe.

However I have refactored the code to avoid duplicating it. Let me know what you think.

File: src/sqlfluff/rules/L057.py
Comment (tunetheweb) [reply src/sqlfluff/rules/L057.py]: I've gone with `additional_allowed_characters` as it's a rule-specific config specific to this rule, so think it's obvious it's for identifiers.

**Review by barrywhart — APPROVED** [review]
Looks good! Thanks for the refactoring -- it is 💯.

**Comment by tunetheweb** [comment]
> Looks good! Thanks for the refactoring -- it is 💯.

I had another go to make it even cleaner. WDYT?

**Comment by barrywhart** [comment]
New version looks good, also. When I work on rule code, I am often unsure whether it's better to have one long function or several short ones. Lately, I tend to prefer several short ones, if it helps divide the code into logical sections. I think that makes it easier for maintainers, as it potentially lets them focus on the relevant area of the code without having to worry about the whole function. 🤷‍♂️

**Comment by tunetheweb** [comment]
Yeah it's a judgement call.

I think short ones are good if we want to abstract away a specific bit of function. With this one here, we were basically abstracting away one line calls so having a function seemed overkill. Also a near-identical `replace` calls was left behind in main function (the BigQuery bit), so again didn't seem right to separate them. Not to mention there was the `if` in the function for the `allow_space_in_identifier` which seemed to pass a parameter for the sake of something that was better handled in the one `if` statement, rather than two.

All in all, I think this is better as just one function as pretty short (and a lot of comments, without which it would be shorter still). But any larger, I agree might want to split it up.

**Pull request merged**
